### PR TITLE
add anounce- and (pre/|post)RevokeUser signals for non-native backends

### PR DESCRIPTION
**Pull request by blizzz**
Reasoning: in some cases it is useful to listen and react on "new" users. Currently, there are two signals in this context `preCreateUser` and `postCreateUser`. In both cases passwords are sent out to the listeners. This works well for the native backend, but non-native backends typically have no knowledge about the password. Using the very same hook without a password, would create problems with implementations that rely on the password given (one known would be encryption). 

For removal or deletion there are `(pre|post)DeleteUser` which could be re-used quite fine. However, they are typically used to delete data. Unfortunately there are scenarios in practice (against recommendations) where mappings are being cleared only to be restored later. The new `(pre|post)RevokeUser` signals are a soft version of the delete signals that can be used for some light cleanup (e.g. indices like local addressbook) which is easily filled again when users are announced while no user data shall be removed. It seems somewhat unnecessary and of duplicate effort, the goal is to be careful and conservative and avoid unintended data loss – even if the usage is intended. 

A way to test it:

1. have user_ldap configured
2. have admin_audit enabled
3. browser users page and/or delete user mapping (ldap expert settings)
→ you should see entries like:

```
{… "app":"admin_audit","method":"","url":"--","message":"UserID assgined: \"person_62692@example.org\"" …}
{… "app":"admin_audit","method":"POST","url":"\/master\/index.php\/apps\/user_ldap\/ajax\/clearMappings.php","message":"UserID unassigned: \"person_28671@example.org\""… }
```

### CODE REVIEW

**Comment by nickvergessen** [comment]
Are announce/revoke LDAP terms? they sound very random to me and it is not obvious that one is the revert of the other

**Comment by MorrisJobke** [comment]
Also we should document all those events properly on one page in the documentation.

**Comment by blizzz** [comment]
> Are announce/revoke LDAP terms? they sound very random to me and it is not obvious that one is the revert of the other

no, they aren't. announce does not have a direct antonym. Perhaps we go for "assignedUserId" and "unassignedUserId" if that helps? Better ideas welcome.

**Comment by nickvergessen** [comment]
> no, they aren't. announce does not have a direct antonym. Perhaps we go for "assignedUserId" and "unassignedUserId" if that helps? Better ideas welcome.

Sounds good enough

**Comment by blizzz** [comment]
signal names renamed, and fixed tests by DIing the server's user manger.

File: apps/admin_audit/lib/Actions/UserManagement.php [hunk 1]
```diff
@@ -50,6 +50,19 @@ public function create(array $params) {
...
 	}
 
+	/**
+	 * Log assignments of users (typically user backends)
+	 *
+	 * @param string $uid
+	 */
+	public function announce(string $uid) {
+		$this->log(
+		'UserID assgined: "%s"',
```
Comment (nickvergessen) [inline apps/admin_audit/lib/Actions/UserManagement.php]: typo

File: apps/admin_audit/lib/AppInfo/Application.php [hunk 2]
```diff
@@ -93,6 +93,8 @@ protected function userManagementHooks(ILogger $logger) {
 		/** @var IUserSession|Session $userSession */
 		$userSession = $this->getContainer()->getServer()->getUserSession();
 		$userSession->listen('\OC\User', 'postSetPassword', [$userActions, 'setPassword']);
+		$userSession->listen('\OC\User', 'assignedUserId', [$userActions, 'announce']);
+		$userSession->listen('\OC\User', 'postUnassignedUserId', [$userActions, 'revoke']);
```
Comment (nickvergessen) [inline apps/admin_audit/lib/AppInfo/Application.php]: fix method names to new signal names?

**Review by nickvergessen — APPROVED** [review]
fine with me, typo must be fixed, method name nice to have

**Comment by blizzz** [comment]
@nickvergessen done and rebased

File: apps/dav/lib/HookManager.php [hunk 3]
```diff
@@ -77,14 +77,22 @@ public function setup() {
 			'post_createUser',
 			$this,
 			'postCreateUser');
+		\OC::$server->getUserManager()->listen('\OC\User', 'assignedUserId', function ($uid) {
+			$this->postCreateUser(['uid' => $uid]);
```
Comment (MorrisJobke) [inline apps/dav/lib/HookManager.php]: Is `$this` even available here? Maybe assign it on the outside and then use something like `function($uid) use ($outerThis) {` or am I wrong here?

Comment (blizzz) [reply apps/dav/lib/HookManager.php]: >  As of PHP 5.4.0, when declared in the context of a class, the current class is automatically bound to it, making $this available inside of the function's scope. If this automatic binding of the current class is not wanted, then static anonymous functions may be used instead. 

http://php.net/manual/en/functions.anonymous.php

**Review by MorrisJobke — APPROVED** [review]
🐘

**Pull request merged**